feat: formatting, fixes and improvements

paulbuechner opened the pull request:
Hi Mike,

appreciate your work in providing some maintenance love to the project.

This PR includes the following features, fixes, and improvements:

- Formatting by: https://github.com/sumneko/lua-language-server/wiki/Formatter
- Updated `BlizzardActionBars` to include all available ActionBars:
    - #177
- Temporarily disable MainMenuBar/MainActionBar tooltip to prevent: `[ADDON_ACTION_BLOCKED] AddOn 'MegaMacro' tried to call the protected function ''`
    - #200
- Temporarily disable range-based hotkey coloring to prevent:
    - #176 
- UI fixes and improvements
- Refactor EventFrame to support fixed update intervals
- Add linting support for new `known` and `unknown` conditionals:
     - #199
- Add UTF-8 support when parsing characters 
- Fix `channeling`/`nochanneling` modifiers
- Updated TOC
- Updated README

Cheers 😊

fixes #177, #199

paulbuechner committed:
chore: formatting

paulbuechner committed:
chore: fix BlizzardActionBars

paulbuechner committed:
chore: update TOC

paulbuechner committed:
chore: minor improvements

chore: update vscode globals recognition

fix(formatting): unset max line length

- remove tab indents in favor of spaces

chore: update .editorconfig

paulbuechner committed:
chore: replace `Action` with `MainMenuBar`

- fixes #131, #173, #172

paulbuechner committed:
feat(tooltip): use native GameTooltip for config checkbuttons

- enables reduced tooltip fontsize

paulbuechner committed:
fix(ui): minor improvements

- fix `Mega Macro` title to no longer get hidden behind top frame bar
- adjust size of `Save` and `Cancel` buttons (grow buttons to make use of unused whitespace between them)
- fix `FramePortrait` placement
- fix `Lock/Unlock` button size to correctly match frame border height
- update xml intentation to use spaces (be consistent across other files)

paulbuechner commented:
Do not merge yet... more fixes are coming. Preferably convert it to draft.

Also, I'd love some feedback on this PR, so we can discuss the changes I made.  😊

paulbuechner committed:
feat(libs): add support for utf-8

paulbuechner committed:
feat(engine): add utf-8 support when parsing characters

*feat(conditionals): add linting support for new `known` and `unknown` conditionals (#199)
*refactor: EventFrame to support fixed update intervalls
*fix: temporarily disable range-based hotkey coloring (#176)
*fix: temporarily disable MainMenuBar (ActionBar) tooltips (#131, #172, #173)
*chore: minor improvements
*ci: version bump

paulbuechner committed:
chore: minor UI fixes

paulbuechner committed:
chore: update README

paulbuechner committed:
chore: update README

paulbuechner committed:
chore: update README

paulbuechner committed:
chore: fix typos

paulbuechner committed:
feat: support space separated string literals in `known`, `unknown` conditionals

*fix: use UTF-8 strlen to determine wordlength

fixes: #199

paulbuechner committed:
chore: more utf8 adjustments

aurelion314 commented:
FYI something is causing some stutter with this branch. I also got some errors during solo shuffle but was too into it to remember what it was heh. I'll keep playing with it but if you have any idea what is causing the CPU usage let me know.

For reference, here is AddonUsage for both version. This branch is using more than twice the CPU, and almost 20mb more memory:
![image](https://user-images.githubusercontent.com/23002984/225562701-a66ab5cc-f5dd-4241-9346-7d640c733c08.png)

Current version:
![image](https://user-images.githubusercontent.com/23002984/225562926-7ec03a98-d4f3-4857-b452-7ceabb0e3f2d.png)

paulbuechner commented:
Thanks for the feedback. I'll definitely have a look soon 👀

aurelion314 commented:
It seems to be related to UTF-8, even though the error I saw actually said it came from WeakAura, but it doesn't do that when I revert to the old version, so I think the UTF-8 change here is being shared and not playing nice with other addons.

JoeLatte88 commented:
has anyone had time to dive deeper into the potential UTF-8 issue? I may take a peek to see if I can help, but my exp with these things are very limited.

aurelion314 commented:
Maybe we can turn this into smaller PRs? @paulbuechner or @JoeLatte88. It is difficult to dig into what actually changed since so many files have formatting updates. Combine that with multiple feature changes and it becomes difficult to isolate what is causing the bug and/or the increased CPU load.

Unfortunately I've been busy and not playing WoW much so I am unlikely to dig through this myself in the near future :\

JoeLatte88 commented:
I agree on the "smaller PRs"

JoeLatte88 commented:
@paulbuechner , perhaps we should roll back the UTF-8 changes and deliver the problem fixes for the time being. I'll go ahead and cherry pick UTF-8 changes out in a clone of your branch and verify if performance is back to what we'd expect. At least then, we can push this out to fix issues and then work on optimization for UTF-8

paulbuechner commented:
Hey! I'm on it today... let me check some changes I made to confirm if the issue is really related to UTF-8.

paulbuechner commented:
@aurelion314 @JoeLatte88 

Some information I've found:

The massive memory consumption is caused when loading the icons [here](https://github.com/paulbuechner/mega-macro/blob/master/src/main.lua#L9) to make them searchable in the icon browser when creating a macro.

If we disable the feature the icon starts idling around 700k
![grafik](https://user-images.githubusercontent.com/45827409/227784959-edfb85db-c9fd-429d-8930-842d1a686de7.png)

Nevertheless, a memory leak/creep is slowly ticking up...

![grafik](https://user-images.githubusercontent.com/45827409/227785160-8fe24ca9-043a-4e81-afba-0fc2b7ddbf30.png)

EDIT: Reading some resources on this topic I think we should be fine with this issue:
![grafik](https://user-images.githubusercontent.com/45827409/227798870-ae8a57fb-6f66-4d0b-bbc1-c5da28f35268.png)

Regarding the massive memory consumption... I propose rewriting the icon browser to only load when creating a macro. Maybe something like WeakAura has, which works like:

1. Do not load & show any icons in the icon browser until the user enters the spell name. This will prevent us from loading all icons simultaneously -> only load and cache what is requested.
2. Implement something like [use-debounce](https://github.com/xnimorz/use-debounce) to make loading the requested icons, even more, memory/CPU efficient.
3. Invalidate cache if icon browser is closed / macro is created to free up memory.

paulbuechner committed:
chore(style): use consistent quotes "double"

paulbuechner committed:
fix: only update actionbar on frame render

paulbuechner committed:
chore: utf-8 adjustments

paulbuechner committed:
fix: add null check for actionbar iterator

paulbuechner committed:
chore: update toc

paulbuechner commented on src/main.lua:
```diff
@@ -9,16 +9,16 @@ local function MegaMacroOnUpdate(self, elapsed)
-
-        self.TimeSinceLastUpdate = self.TimeSinceLastUpdate - MegaMacroUpdateInterval;
-    end
+    -- Limit update rate to fixed interval
+    -- self.TimeSinceLastUpdate = self.TimeSinceLastUpdate + elapsed;
+    -- while (self.TimeSinceLastUpdate > MegaMacroUpdateInterval) do
+    --     self.TimeSinceLastUpdate = self.TimeSinceLastUpdate - MegaMacroUpdateInterval;
+    -- end
+    MegaMacroActionBarEngine.Update(elapsed)
```
Regarding the additional CPU consumption: This may have been caused by forcing an action bar update even if no new frame has been rendered yet. I reverted this change to only update on frame render.

paulbuechner committed:
fix(parsing): `channeling`/`nochanneling` modifier

- allow for OptionalWordModifier using channeling
- add nochanneling conditional
- remove channelling conditional (typo)

JoeLatte88 commented:
@paulbuechner , is the current code is a safe state from a performance perspective for us to review for approval? this is already a pretty big PR hehehe :)

would make sense to push and release what we have and we can work on further optimizations and such in a separate PR to at least get the fixes out

paulbuechner commented:
The current state works well for me. At the moment I've no further changes planned. I gave you the info I figured out about @aurelion314 concerns here if you would like to do further investigations.

Also I've documented any noticeable changes at the top.

JoeLatte88 commented:
@aurelion314 , I reviewed the code and a VAST majority of the files changed are just adding spaces or updating tabbed spacing across some files. I don't see an issue with some of the updates from a code perspective which aren't already mentioned in @paulbuechner notes which will go as part of the addon release notes.

aurelion314 commented:
Nice! Unfortunately I stopped playing and don't even have WoW installed to try this out >_>

@paulbuechner do you want to make this PR directly into the original (https://github.com/Sellorio/mega-macro)? Sellorio has to finalize the merge anyway, and I don't want to hold up your updates.

paulbuechner commented:
Ok. I opened the PR (https://github.com/Sellorio/mega-macro/pull/202).

Let's see how it goes 😃

paulbuechner committed:
chore: begin migration to new `ScrollFrame` API

- TODO: `IconBrowser` migration

paulbuechner committed:
feat: add `IconTexture` to display in addon explorer
